src/highdicom/seg/sop.py

```diff
@@ -726,29 +742,48 @@ def add_segments(
             for index in range(plane_position_values.shape[1])
         ]
 
-        # When using binary segmentation type, the previous frames may have been
-        # padded to be a multiple of 8. In this case, we need to decode the
-        # pixel data, add the new pixels and then re-encode. This process
-        # should be avoided if it is not necessary in order to improve
-        # efficiency.
-        if (self.SegmentationType == SegmentationTypeValues.BINARY.value and
-                ((self.Rows * self.Columns * self.SamplesPerPixel) % 8) > 0):
-            re_encode_pixel_data = True
-            logger.warning(
-                'pixel data needs to be re-encoded for binary bitpacking - '
-                'consider using FRACTIONAL instead of BINARY segmentation type'
-            )
-            # If this is the first segment added, the pixel array is empty
-            if hasattr(self, 'PixelData') and len(self.PixelData) > 0:
-                full_pixel_array = self.pixel_array.flatten()
-            else:
-                full_pixel_array = np.array([], np.bool)
-        else:
-            re_encode_pixel_data = False
-
+        # In certain circumstances, we can add new pixels without unpacking the
+        # previous ones, which is more efficient. This can be done when using
+        # non-encapsulated transfer syntaxes when there is no padding required
+        # for each frame to be a multiple of 8 bits.
+        framewise_encoding = False
+        is_encaps = self.file_meta.TransferSyntaxUID.is_encapsulated
+        if not is_encaps:
+            if self.SegmentationType == SegmentationTypeValues.FRACTIONAL.value:
+                framewise_encoding = True
+            elif self.SegmentationType == SegmentationTypeValues.BINARY.value:
+                # Framewise encoding can only be used if there is no padding
+                # This requires the number of pixels in each frame to be
+                # multiple of 8
+                if (self.Rows * self.Columns * self.SamplesPerPixel) % 8 == 0:
+                    framewise_encoding = True
+
+        if framewise_encoding:
             # Before adding new pixel data, remove trailing null padding byte
             if len(self.PixelData) == get_expected_length(self) + 1:
                 self.PixelData = self.PixelData[:-1]
+        else:
+            if self.SegmentationType == SegmentationTypeValues.BINARY.value:
+                if not is_encaps:
```

Comment:
There is a check above asserting that the Pixel Data element is not encapsulated in case of `SegmentationTypeValues.BINARY`

Reply:
Ah yes, moved this warning to a logically equivalent but more convenient place a few lines above